### ISSUE TITLE
Not last return javadoc tag cause parse error

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#12864]
I have read check documentation: https://checkstyle.org/config_javadoc.html#NonEmptyAtclauseDescription
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

detected at https://github.com/checkstyle/checkstyle/pull/12862#issuecomment-1474850368

```bash
rivanov@p5510:/var/tmp$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name = "Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
       <module name="NonEmptyAtclauseDescription"/>
    </module>

</module>
rivanov@p5510:/var/tmp$ cat Test.java 
class Test {
  /**
  * Some summary.
  * @return
  * @param a Some description
  */
  public int method(String a, int b) throws Exception
  {
    return 1;
  }
}
rivanov@p5510:/var/tmp$ java -jar checkstyle-10.9.1-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:5: Javadoc comment at column 0 has parse error. Details: no viable alternative at input '  *' while parsing JAVADOC_TAG [NonEmptyAtclauseDescription]
Audit done.
Checkstyle ends with 1 errors.
```

Expected:
to have no parse error as we do then return is last tag
```
rivanov@p5510:/var/tmp$ cat Test.java 
class Test {
  /**
  * Some summary.
  * @param a Some description
  * @return
  */
  public int method(String a, int b) throws Exception
  {
    return 1;
  }
}
rivanov@p5510:/var/tmp$ java -jar checkstyle-10.9.1-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:5: At-clause should have a non-empty description. [NonEmptyAtclauseDescription]
Audit done.
Checkstyle ends with 1 errors.
```

**Comment by romani** [comment]
Avoid issues with antrl label, and try deal with first/second/third gsoc issues.

**Comment by nrmancuso** [comment]
@romani why do we need check to create issue on parsing problems?

```
➜  src cat Test.java
class Test {
  /**
  * Some summary.
  * @param a Some description
  * @return
  */
  public int method(String a, int b) throws Exception
  {
    return 1;
  }
}

➜  src java -jar checkstyle-10.7.0-all.jar -J Test.java
COMPILATION_UNIT -> COMPILATION_UNIT [1:0]
...
        |   |   |   |   `--JAVADOC -> JAVADOC [2:5]
...
        |   |   |   |       |--WS ->   [5:3]
        |   |   |   |       |--JAVADOC_TAG -> JAVADOC_TAG [5:4]
        |   |   |   |       |   |--RETURN_LITERAL -> @return [5:4]
        |   |   |   |       |   |--NEWLINE -> \n [5:11]
        |   |   |   |       |   `--DESCRIPTION -> DESCRIPTION [6:0]
        |   |   |   |       |       `--TEXT ->    [6:0]
        |   |   |   |       `--EOF -> <EOF> [6:2]
...

➜  src cat Test.java 
class Test {
  /**
  * Some summary.
  * @return
  * @param a Some description
  */
  public int method(String a, int b) throws Exception
  {
    return 1;
  }
}


➜  src java -jar checkstyle-10.7.0-all.jar -J Test.java
Exception in thread "main" java.lang.IllegalArgumentException: [ERROR:5] Javadoc comment at column 0 has parse error. Details: no viable alternative at input '  *' while parsing JAVADOC_TAG
        at com.puppycrawl.tools.checkstyle.DetailNodeTreeStringPrinter.parseJavadocAsDetailNode(DetailNodeTreeStringPrinter.java:70)
        at com.puppycrawl.tools.checkstyle.AstTreeStringPrinter.parseAndPrintJavadocTree(AstTreeStringPrinter.java:116)
...


```

**Comment by romani** [comment]
This is searchable example how our users might encounter this problem.
Copy paste of ready config and input from investigation in another issue.

**Comment by 0xbakry** [comment]
@romani I'm trying to solve this issue and i think it's not about **return** position, since something like: 
```
class Test {
  /**
  * Some summary.
  * @return
  * anything OR blank
  * @param a Some description
  */
  public int method(String a, int b) throws Exception
  {
    return 1;
  }
}
``` 
will work correctly.

I debugged the code and find that the exception is thrown from `getInterpreter().adaptivePredict` function in `JavadocParser.java`.
and I think this happens cause of `decision` variable, since it tries to retrieve the `DFA` according to `decision` and since the automaton doesn't have a way for any `LITERAL` it throws this error.
Please correct me if I'm wrong

**Comment by rnveach** [comment]
The issue is most likely because the grammar says the description is not optional.
https://github.com/checkstyle/checkstyle/blob/1346ff94acef40b7772b5170beeb5fbc5d5b7d5a/src/main/resources/com/puppycrawl/tools/checkstyle/grammar/javadoc/JavadocParser.g4#L1150
It would need a `?` to be optional. Since there is none, it is required to exist.

@0xbakry 's example is seen as a description since it is just plain text after the return.
Oddly, the following example is seen as valid even though there is technically no description.
````
  * @param a Some description
  * @return
  */
````
The description is picked up as the spaces right before `*/`.

Our javadoc grammar has issues overall and probably needs a rewrite.

**Comment by 0xbakry** [comment]
@rnveach I was thinking of that the description must be provided, but 
``` 
  * @param a Some description
  * @return
  */
```
confused me a little so I dismissed this assumption, and you clear it for me now, thanks.

> The description is picked up as the spaces right before */.

But I think it take the `*/` itself as a description, look:
https://github.com/checkstyle/checkstyle/blob/1346ff94acef40b7772b5170beeb5fbc5d5b7d5a/src/main/resources/com/puppycrawl/tools/checkstyle/grammar/javadoc/JavadocParser.g4#L1126

**Comment by rnveach** [comment]
> But I think it take the */ itself as a description

`*/` is the end of the comment/javadoc and is not given to the Javadoc parser.

If you print the tree with javadoc you will see:
````
        |  |  |  |      `--EOF -> <EOF> [6:2]       <---- End of Javadoc Tokens
        |  |  |  `--BLOCK_COMMENT_END -> */ [6:2]   <---- Back to Java Tokens
````

**Comment by romani** [comment]
@0xbakry , 

>I'm trying to solve this issue

Be careful, all antrl issues are easy if you are new to project

**Comment by 0xbakry** [comment]
I know it's not that easy and that's why I picked it, just to go deeper in the project and I hope that I'm not doing wrong (:

For now, I updated the grammar as @rnveach said and added new tests to suit with the new grammar but I get `Rule violated for class com.puppycrawl.tools.checkstyle.checks.javadoc.NonEmptyAtclauseDescriptionCheck: branches covered ratio is 0.75, but expected minimum is 1.00` 

I created the `pit-reports` and I'm still trying to figure out what I have missed, so my question is that:
Do I have to do the tests somewhere other than `com.puppycrawl.tools.checkstyle.checks.javadoc.nonemptyatclausedescription` ?
Just to know If I'm trying in the correct place. 
Thanks

**Comment by rnveach** [comment]
Error is not from pitest. It is basic code coverage of jacoco. I don't know what changes you made, and the best place to view your changes is in a PR. I can only assume here.

Besides fixing the issue, you will need to add new tests and inputs that show proof that the problem is resolved and works in different scenarios where possible. Your new tests may be in `JavadocParseTreeTest`. The issue also mentions `NonEmptyAtclauseDescription`, so a new test should be in `NonEmptyAtclauseDescriptionCheckTest` to show the expected behavior. Other tests may be needed depending on CI runs or from reviewers.

I didn't think any new changes would be in production code besides the grammar, but it depends on what is needed and what you have done.

**Comment by nrmancuso** [comment]
@0xbakry I strongly suggest that you focus on only grammar update and send PR for that before spending time updating checks (if that's even necessary). Though this issue was reported with a check, it is not necessary to use check to reproduce and fix the problem, see https://github.com/checkstyle/checkstyle/issues/12864#issuecomment-1475058784.

Please send initial draft PR with only the following so that we can verify approach:
1. grammar update
2. new test/test inputs for JavadocParseTreeTest

We have some performance issues with Javadoc grammar, so not only must solution be syntactically valid, but performant as well.

**Comment by nrmancuso** [comment]
Closed via #12979

**Comment by nrmancuso** [comment]
@0xbakry thanks a lot for this contribution!
